packages/gen-ai-hub/src/utils/deployment-cache.ts

```diff
@@ -0,0 +1,83 @@
+// TODO: should we move this to @sap-cloud-sdk/connectivity? or add some kind of exception to the rule if possible?
+// eslint-disable-next-line import/no-internal-modules
+import { Cache } from '@sap-cloud-sdk/connectivity/internal.js';
+import { type AiDeployment } from '@sap-ai-sdk/ai-core';
+import { type DeploymentResolutionOptions } from './deployment-resolver.js';
+import { extractModel, type FoundationModel } from './model.js';
+
+function getCacheKey({
+  scenarioId,
+  executableId,
+  model,
+  resourceGroup = 'default'
+}: DeploymentResolutionOptions) {
+  return `${scenarioId}-${executableId ?? ''}-${model?.name ?? ''}-${model?.version ?? ''}-${resourceGroup}`;
+}
+
+interface Deployment {
+  id: string;
+  model?: FoundationModel;
+}
+
+/**
+ * Create a cache for deployments.
+ * @param cache - Pure cache object.
+ * @returns The deployment cache.
+ * @internal
+ */
+function createDeploymentCache(cache: Cache<Deployment>) {
+  return {
+    /**
+     * Get a deployment from the cache.
+     * @param opts - Deployment resolution options to get the cached deployment for.
+     * @returns The cached deployment or undefined if not found.
+     */
+    get: (opts: DeploymentResolutionOptions): Deployment | undefined =>
+      cache.get(getCacheKey(opts)),
+    /**
+     * Store a deployment in the cache.
+     * @param opts - Deployment resolution options to set the deployment for.
+     * @param deployment - Deployment to cache.
+     */
+    set: (
+      opts: DeploymentResolutionOptions,
+      deployment: AiDeployment
+    ): void => {
+      cache.set(getCacheKey(opts), {
+        entry: transformDeploymentForCache(deployment)
+      });
+    },
+    /**
+     * Store multiple deployments in the cache, based on the model from the respective AI deployments.
+     * @param opts - Deployment resolution options to set the deployments for. Model information in the deployment resolution options are ignored.
+     * @param deployments - Deployments to retrieve the IDs and models from.
+     */
+    setAll: (
+      opts: Omit<DeploymentResolutionOptions, 'model'>,
+      deployments: AiDeployment[]
+    ): void => {
+      // go backwards to cache the first deployment ID for each model
+      deployments.reverse().forEach(deployment => {
+        cache.set(getCacheKey({ ...opts, model: extractModel(deployment) }), {
+          entry: transformDeploymentForCache(deployment)
+        });
+      });
+    },
+    clear: () => cache.clear()
+  };
+}
+
+function transformDeploymentForCache(deployment: AiDeployment): Deployment {
+  return {
+    id: deployment.id,
+    model: extractModel(deployment)
+  };
+}
+
+/**
+ * Cache for deployments.
+ * @internal
+ */
+export const deploymentIdCache = createDeploymentCache(
+  new Cache(5 * 60 * 1000) // 5 minutes
```

Comment:
What is our invalidation strategy here?

I remember for the Cloud SDK we used the auth token lifetime, e.g.:
```ts
  const expiresIn = first(destination.authTokens || [])?.expiresIn;
  const expirationTime = expiresIn
    ? Date.now() + parseInt(expiresIn) * 1000
    : undefined;
 ```
 If I understand the PR correctly, if a deployment is removed, we would send failing request for up to 5 minutes?

Reply:
yes

Reply:
5 minutes feels rather arbitrary

Reply:
I do see a `cache.clear(`), so for testing its still fine

Reply:
I will check how realistic it is to implement this. I think it generally makes sense, however the more flexibility the user has, the more difficult it will be to decide whether a 404 is based on an incorrect deployment id or something else.

Reply:
we agreed that we will do this as a follow if possible

Reply:
I think it's unlikely that deployments will change often.

* Any new models will still be picked up, regardless of cache duration.
* Changing the deployment of an existing model has no benefits and I wouldn't know a use case for doing that
* Deleting a deployment can be done, but requests would fail regardless of caching or not, because the desired model no longer exists. The failures would be slightly more explicit without the cache, but the 404 should also be somewhat clear, in particular with proper logging 😉 
* For local testing caching is usually not a problem

In short: I think we should cache quite long by default, e.g. 12hrs or even more